sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Analyzer.scala

```diff
@@ -1918,14 +1918,9 @@ class Analyzer(
       case p: Project => p
       case f: Filter => f
 
-      // todo: It's hard to write a general rule to pull out nondeterministic expressions
-      // from LogicalPlan, currently we only do it for UnaryNode which has same output
-      // schema with its child.
-      case p: UnaryNode if p.output == p.child.output && p.expressions.exists(!_.deterministic) =>
-        val nondeterministicExprs = p.expressions.filterNot(_.deterministic).flatMap { expr =>
-          val leafNondeterministic = expr.collect {
-            case n: Nondeterministic => n
-          }
+      case p: UnaryNode if p.expressions.exists(!_.deterministic) =>
```

Comment:
It looks like the ones that are interesting are:

- Pivot
- Window
- Aggregate
- RedistributeData (RepartitionByExpression and SortPartitions)
- Sort